summary-sheets/02-problem-solving-paradigms.md

```diff
@@ -0,0 +1,388 @@
+       for (int i=0; i<m; i++) 
+       { 
+         if (coins[i] <= V) 
+         { 
+             int sub_res = minCoins(coins, m, V-coins[i]); 
+       
+             // Check for INT_MAX to avoid overflow and see if 
+             // result can minimized 
+             if (sub_res != Integer.MAX_VALUE && sub_res + 1 < res) 
+                res = sub_res + 1; 
+         } 
+       } 
+       return res; 
+    } 
+    public static void main(String args[]) 
+    { 
+       int coins[] =  {9, 6, 5, 1}; 
+       int m = coins.length; 
+       int V = 11; 
+       System.out.println("Minimum coins required is "+ minCoins(coins, m, V) ); 
+    } 
+
+}/* This code is contributed by Rajat Mishra */
+~~~
+
+**Output**
+~~~
+Minimum coins required is 2
+~~~
+
+If we draw the recursive tree, you can see that many subrpoblems are solved more than once. For example, 
+when we start from V = 11, we can reach 6 by subtracting one 5 times and by subtracting 5 one times. 
+So the subproblem for 6 is called twice.
+
+In other words, this problem has Overlapping Subprolems property. We can use memoization to avoid 
+recomputing the overlapping subproblems. This can be done by constructing a temporary array table[][] in 
+bottom up manner. Below is Dynamic Programming based solution.
+
+~~~java
+// A Dynamic Programming based Java 
+// program to find minimum of coins 
+// to make a given change V 
+
+import java.io.*; 
+
+  
+
+class GFG  
+{ 
+    // m is size of coins array  
+    // (number of different coins) 
+    static int minCoins(int coins[], int m, int V) 
+    { 
+        // table[i] will be storing  
+        // the minimum number of coins 
+        // required for i value. So  
+        // table[V] will have result 
+        int table[] = new int[V + 1]; 
+        
+        // Base case (If given value V is 0) 
+        table[0] = 0; 
+
+        // Initialize all table values as Infinite 
+        for (int i = 1; i <= V; i++) 
+        table[i] = Integer.MAX_VALUE; 
+
+        // Compute minimum coins required for all 
+        // values from 1 to V 
+
+        for (int i = 1; i <= V; i++) 
+        { 
+            // Go through all coins smaller than i 
+            for (int j = 0; j < m; j++) 
+            if (coins[j] <= i) 
+            { 
+                int sub_res = table[i - coins[j]]; 
+                if (sub_res != Integer.MAX_VALUE  
+                       && sub_res + 1 < table[i]) 
+                       table[i] = sub_res + 1; 
+            } 
+        } 
+        return table[V]; 
+    } 
+    // Driver program  
+    public static void main (String[] args)  
+    { 
+        int coins[] = {9, 6, 5, 1}; 
+        int m = coins.length; 
+        int V = 11; 
+        System.out.println ( "Minimum coins required is " 
+                            + minCoins(coins, m, V)); 
+    } 
+} 
+
+//This article is contributed by vt_m. 
+~~~
+
+**Output**
+~~~
+Minimum coins required is 2
+~~~
+

```

Comment:
Awesome job! Really good stuff, in the future let's try to bullet point stuff as much as possible for faster reading, but this is dope!